src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -274,60 +274,80 @@ bool AbstractLQPNode::has_matching_unique_constraint(const ExpressionUnorderedSe
 }
 
 std::vector<FunctionalDependency> AbstractLQPNode::functional_dependencies() const {
-  // Gather input FDs from all input nodes
-  auto fds_left = std::vector<FunctionalDependency>();
-  auto fds_right = std::vector<FunctionalDependency>();
-  if (left_input()) {
-    fds_left = left_input()->functional_dependencies();
-  }
-  if (right_input()) {
-    fds_right = right_input()->functional_dependencies();
+  // (1) Gather FDs from previous nodes
+  auto fds_in = pass_functional_dependencies();
+  if constexpr (HYRISE_DEBUG) {
+    auto distinct_fds = std::vector<FunctionalDependency>();
+    for (const auto& fd : fds_in) {
+      Assert(std::none_of(distinct_fds.cbegin(), distinct_fds.cend(),
+                          [&fd](const auto& distinct_fd) { return fd == distinct_fd; }),
+             "Did not expect LQP nodes to pass duplicate FDs.");
+      distinct_fds.push_back(fd);
+    }
   }
 
-  if (HYRISE_DEBUG && !fds_right.empty()) {
-    for (const auto& fd_right : fds_right) {
-      const bool duplicate = std::any_of(fds_left.begin(), fds_left.end(),
-                                         [&fd_right](const auto& fd_left) { return (fd_left == fd_right); });
-      Assert(!duplicate,
-             "Unexpected duplicate functional dependency found in " + this->description(DescriptionMode::Short));
+  // (2) Generate FDs from node's unique constraints
+  const auto& output_expressions = this->output_expressions();
+  const auto unique_constraints = this->unique_constraints();
+  for (const auto& unique_constraint : *unique_constraints) {
+    auto determinants = unique_constraint.expressions;
+    auto dependents = ExpressionUnorderedSet();
+    for (const auto& output_expression : output_expressions) {
+      if (!determinants.contains(output_expression)) {
+        dependents.insert(output_expression);
+      }
+    }
+
+    // Add the FD to the output if not already existing
+    if (!dependents.empty() &&
+        std::find_if(fds_in.cbegin(), fds_in.cend(), [&determinants, &dependents](const auto& fd) {
+          return (fd.determinants == determinants) && (fd.dependents == dependents);
+        }) == fds_in.cend()) {
+      fds_in.emplace_back(determinants, dependents);
     }
   }
 
-  auto& fds_in = fds_left;
-  fds_in.insert(fds_in.end(), fds_right.cbegin(), fds_right.cend());
-
-  // Currently, we do not support FDs in conjunction with null values in their determinant set.
-  // Some LQP nodes, however, change column nullability (like for instance outer joins). Therefore, we check input FDs
-  // for compliance and discard them, if necessary.
+  /**
+   * (3) Remove invalid FDs:
```

Comment:
This only applies to the FDs created from (2), correct?

Reply:
No, this is a general rule. It also applies for FDs from (1).

Reply:
But a column should not suddenly become nullable. If an FD from (1) acts on a nullable column, isn't the input faulty?

Reply:
Yes, it is. Since `pass_functional_dependencies()` is a pure helper function to pass FDs, it is not required to filter FDs. It is only required to check for duplicates before adding to it.

By design, validity-checks are performed at the end, in the top node.

Reply:
Mmmh, I'm not sure if I like this. For me, it would be more natural to (1) debug-assert that FDs coming from `pass_functional_dependencies` are not based on nullable columns and (2) skip the FD generations in the `for (const auto& output_expression : output_expressions)` block if the expression is nullable.

I won't block the PR because of this, but please consider this.